### PR TITLE
Temp file handling

### PR DESCRIPTION
**Pull request by agr**
Added classes to abstract away and handle creation and lifetime of temporary files.

Progress on https://github.com/NuGet/Engineering/issues/1281

### CODE REVIEW

File: src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /// Opens existing file for exclusive read and deletes it on close.
+    /// </summary>
+    public class DeleteOnCloseReadOnlyTempFile : ITempReadOnlyFile
+    {
+        private const int BufferSize = 8192;
+        private readonly FileStream _fileStream;
+
+        public DeleteOnCloseReadOnlyTempFile(string fileName)
+        {
+            _fileStream = new FileStream(
```
Comment (joelverhagen) [inline src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: Can we share this logic with `FileStreamUtility`?

File: src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+                fileName,
+                FileMode.Open,
+                FileAccess.Read,
+                FileShare.None,
+                BufferSize,
+                FileOptions.DeleteOnClose | FileOptions.Asynchronous);
+        }
+
+        public async Task<string> ReadToEndAsync()
```
Comment (joelverhagen) [inline src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: How big will these files be? Is it really appropriate for these to be entirely in memory?

Comment (agr) [reply src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: Less than 1 KB in my case.

File: src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs (hunk 1 above)
Comment (agr) [reply src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: Done.

File: src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        public DeleteOnCloseReadOnlyTempFile(string fileName)
+        {
+            _fileStream = FileStreamUtility.OpenTemporaryFile(fileName);
+        }
+
+        public async Task<string> ReadToEndAsync()
+        {
+            using (var streamReader = new StreamReader(ReadStream, Encoding.UTF8, detectEncodingFromByteOrderMarks: false, bufferSize: BufferSize, leaveOpen: true))
+            {
+                return await streamReader.ReadToEndAsync();
```
Comment (loic-sharma) [inline src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: >await [](start = 23, length = 5)

You don't need to `await` here, right?

Comment (agr) [reply src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: Not having `await` here provides no benefit that I could come up with and would add confusion if `ReadToEndAsync` throws an exception: `DeleteOnCloseReadOnlyTempFile.ReadToEndAsync` would not be in a stack trace which can complicate figuring out what was called from where.

Comment (chenriksson) [reply src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: The benefit is less allocations, as it wouldn't create an unnecessary closure... though we're not optimizing for memory as-is.

Comment (loic-sharma) [reply src/Validation.Common.Job/TempFiles/DeleteOnCloseReadOnlyTempFile.cs]: Avoiding `await`s when possible decreases pressure on the garbage collector. If avoiding an `await` is easy enough to do, we should do it :)

**Pull request merged**